File: ocaml/difference-of-squares/difference_of_squares.ml
let rec range acc n = 
  match n with 
  0 -> acc |
  _ -> range (n::acc) (n - 1)

let square x =
  x * x

let sum = List.fold_left (+) 0

let square_of_sum x =
  range [] x
  |> sum
  |> square
  
let sum_of_squares x =
  range [] x
  |> List.map square
  |> sum

let difference_of_squares n =
  square_of_sum n - sum_of_squares n

File: cpp/difference-of-squares/difference_of_squares.cpp
#include "difference_of_squares.h"
#include <vector>
#include <iterator>
#include <numeric>
#include <functional>

namespace difference_of_squares {
  int sum_of_squares(int n) {
    std::vector<int> ns(n);
    std::iota(ns.begin(), ns.end(), 1);

    std::vector<int> squares;
    std::transform(ns.begin(), ns.end(), std::back_inserter(squares), 
      [](int x) { return x * x; });

    return std::accumulate(squares.begin(), squares.end(), 0, std::plus<>());
  }

  int square_of_sum(int n) {
    std::vector<int> ns(n);
    std::iota(ns.begin(), ns.end(), 1);

    int sum = std::accumulate(ns.begin(), ns.end(), 0, std::plus<>());

    return sum * sum;
  }

  int difference(int n) {
    return square_of_sum(n) - sum_of_squares(n);
  }
}
